### PR TITLE
create class DocHandler

### PR DESCRIPTION
**Pull request by kasperpeulen**
just refactoring

### CODE REVIEW

File: lib/doc_handler.dart [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import 'dart:math' as math;
+
+import 'dartservices_client/v1.dart';
+import 'editing/editor.dart';
+import 'context.dart';
+import 'services/common.dart';
+import 'src/ga.dart';
+
+import 'dart_pad.dart';
+import 'package:markd/markdown.dart' as markdown;
```
Comment (devoncarew) [inline lib/doc_handler.dart]: nit: sort package: imports before file ones

File: lib/doc_handler.dart [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+class DocHandler {
+  Editor _editor;
+  Context _context;
+  DartservicesApi _dartServices;
```
Comment (devoncarew) [inline lib/doc_handler.dart]: The dart services should be available as `dartServices` from the `dart_pad.dart` import; you won't need to pass it into this classes ctor.

File: lib/doc_handler.dart [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  Analytics ga;
+  static const List cursorKeys = const [KeyCode.LEFT, KeyCode.RIGHT, KeyCode.UP, KeyCode.DOWN];
```
Comment (devoncarew) [inline lib/doc_handler.dart]: nit: let's sort statics before instance fields; also, long line

File: lib/doc_handler.dart [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+  final NodeValidatorBuilder _htmlValidator = new NodeValidatorBuilder.common()
+    ..allowElement('a', attributes: ['href'])
+    ..allowElement('img', attributes: ['src']);
+
+  DocHandler(this._editor, this._context, this._dartServices, this.ga) {
+    keys.bind(['f1'], toggleDocTab);
+    document.onClick.listen((e) => _handleClick(e));
```
Comment (devoncarew) [inline lib/doc_handler.dart]: Right now `playground.dart` is the only file that knows the specifics of the UI. I want to keep that the case for now. It'll let us swap out new UIs easily; we'll only have to write a new top-level file (`playground.dart`, `embedded.dart`, ...). So, some aspects of this file should move back into playground.dart; probably anything that knows about a querySelector call or that operates directly on the results of one.

**Comment by devoncarew** [comment]
I love the idea of splitting code out of playground.dart; it's gotten too big. I'd like to keep code in it that is tied to the specifics of the current UI. So, things that are manipulating specific elements. The documentation code that we can split out should effectively make a library that would be useful to playground.dart - and make that library smaller - but would also work equally well with another envisioning of the dartpad UI, say something named `embedded.dart`. Make sense?

**Comment by kasperpeulen** [comment]
@devoncarew Cool, makes sense. Wil try to put some stuff back in playground.

**Comment by kasperpeulen** [comment]
So I guess the handleClickEvent stuff, also has to go back to
playground.dart ? @devoncarew ?

2015-04-06 0:32 GMT+02:00 Devon Carew <notifications@github.com>:

> I love the idea of splitting code out of playground.dart; it's gotten too
> big. I'd like to keep code in it that is tied to the specifics of the
> current UI. So, things that are manipulating specific elements. The
> documentation code that we can split out should effectively make a library
> that would be useful to playground.dart - and make that library smaller -
> but would also work equally well with another envisioning of the dartpad
> UI, say something named embedded.dart. Make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/dart-pad/pull/264#issuecomment-89866401>.
>



-- 
Kasper

**Comment by devoncarew** [comment]
Yup, @kasperpeulen, I think so. The event handler's could get very short, if they can make use of functionality from the doc_handler library.

File: lib/doc_handler.dart [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  DocHandler(this._editor, this._context, this._ga);
+
+  void generateDoc(DivElement docPanel) {
+    if (!(_context.focusedEditor == 'dart'
+        && _editor.hasFocus
+        && _editor.document.selection.isEmpty)) {
+      return;
+    }
+    _ga.sendEvent('main', 'help');
```
Comment (devoncarew) [inline lib/doc_handler.dart]: Let's move this call to `ga` to playground, when the F1 key is hit. We want to track when the user does something explicitly. Also, it seems odd to send analytics as a side effect of creating a documentation div.

File: lib/doc_handler.dart [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        ..source = _context.dartSource
+        ..offset = offset;
+    }
+
+    // TODO: Show busy.
+    dartServices.document(input).timeout(serviceCallTimeout).then(
+        (DocumentResponse result) {
+      Map info = result.info;
+      String kind = info['kind'];
+      if (info['description'] == null &&
```
Comment (devoncarew) [inline lib/doc_handler.dart]: nit: can fit on one (80-col) line?

File: lib/doc_handler.dart [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+# `${info['description']}`\n\n
+${info['dartdoc'] != null ? info['dartdoc'] + "\n\n" : ""}
+${kind.contains("variable") ? "${info['kind']}\n\n" : ""}
+${kind.contains("variable") ? "**Propagated type:** ${info["propagatedType"]}\n\n" : ""}
+${info['libraryName'] == null ? "" : "**Library:** $apiLink" }\n\n
+''', inlineSyntaxes: [new InlineBracketsColon(), new InlineBrackets()]),
+          validator: _htmlValidator);
+
+        docPanel.querySelectorAll("a").forEach((AnchorElement a)
+        => a.target = "_blank");
```
Comment (devoncarew) [inline lib/doc_handler.dart]: indent an additional 4 chars

File: lib/doc_handler.dart [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        docPanel.querySelectorAll("h1").forEach((h)
+        => h.classes.add("type-${kind.replaceAll(" ","_")}"));
```
Comment (devoncarew) [inline lib/doc_handler.dart]: indent an additional 4 chars

**Comment by devoncarew** [comment]
lgtm! I'll merge once those nits are addressed. Thanks!

**Pull request merged**